timings/state.go

```diff
@@ -0,0 +1,120 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package timings
+
+import (
+	"github.com/snapcore/snapd/overlord/state"
+
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+// timingJson and rootTimingJson aid in marshalling of flattened timings into state.
+type timingJson struct {
+	Level    int           `json:"level"`
+	Label    string        `json:"label,omitempty"`
+	Summary  string        `json:"summary,omitempty"`
+	Duration time.Duration `json:"duration"`
+}
+
+type rootTimingJson struct {
+	timingJson
+	MeasuredSubject string            `json:"subject"`
+	Meta            map[string]string `json:"meta,omitempty"`
+	NestedTimings   []*timingJson     `json:"nested,omitempty"`
+}
+
+var timeDuration = func(start, end time.Time) time.Duration {
+	return end.Sub(start)
+}
+
+// flatten flattens nested measurements into a single list within rootTimingJson.NestedTImings.
+func (t *Timings) flatten() interface{} {
+	data := &rootTimingJson{
+		timingJson: timingJson{
+			Level:    0,
+			Label:    t.m.label,
+			Duration: timeDuration(t.m.stop, t.m.start),
```

Comment:
Shouldn't the parameter order be reversed? WIth the default implementation of `timeDuration`, this will effectively be  `start.Sub(stop)`.

Reply:
Thanks for spotting! Fixed.